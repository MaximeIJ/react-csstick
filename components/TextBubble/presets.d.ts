import { TextBubbleProps } from '../../../../../../../../src/util/types';
export declare const Default: Required<TextBubbleProps>;
