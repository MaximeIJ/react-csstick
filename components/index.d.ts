export { default as Comic, ComicPresets } from './Comic';
export { default as Scene, ScenePresets } from './Scene';
export { default as Stick, StickPresets } from './Stick';
export { default as TextBubble, TextBubblePresets } from './TextBubble';
