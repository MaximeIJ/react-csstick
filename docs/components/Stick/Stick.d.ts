import { StickProps } from '../../../../../../../../src/util/types';
declare const _default: {
    (props: StickProps): JSX.Element;
    displayName: string;
};
export default _default;
