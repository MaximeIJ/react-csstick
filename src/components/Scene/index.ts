export {default} from './Scene';
export * as ScenePresets from './presets';
