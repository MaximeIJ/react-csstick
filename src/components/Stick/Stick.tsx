import React, {FC} from 'react';

import TextBubble from '../TextBubble';

import {Default, Positions} from './presets';

import {chainCall, chainClickable} from '@/hooks/chainClickable';
import './style.css';
import {rotateTransformCSSProp, stickCSSProps} from '@/util/css';
import {LimbAngleProps, StickProps, TextBubbleProps} from '@/util/types';

const Stick: FC<StickProps> = (props = Default) => {
  const {bgColor, color, posId, customPos, dimensions, coord, childProps, children, onClick} = props;
  const hasCustomPose = posId === 'custom' && customPos;
  const {limbs, offsets} = !hasCustomPose ? Positions[posId ?? 'custom'] : customPos;
  const {width, height, thickness} = {...Default.dimensions, ...dimensions};
  const {base = 0, arms, legs} = {...Default.customPos.limbs, ...limbs};

  const stickStyle = stickCSSProps({
    width,
    height,
    thickness,
    bgColor,
    color,
    base,
    offsets,
    coord,
  });

  const renderLimb = (p?: LimbAngleProps) => {
    const limbStyle = p?.angle ? rotateTransformCSSProp(p.angle) : {};
    return <div style={limbStyle} />;
  };

  return (
    <div className="stick" style={stickStyle} onClick={chainCall(onClick)}>
      {/* <div> */}
      <div className="head"></div>
      <div className="body">
        <div className="arms">
          {renderLimb(arms?.left)}
          {renderLimb(arms?.right)}
        </div>
        <div className="legs">
          {renderLimb(legs?.left)}
          {renderLimb(legs?.right)}
        </div>
      </div>
      {(childProps || children)?.map((child, idx) => {
        const {type} = child;
        let ChildComponent = null;
        switch (type) {
          case 'stick':
            ChildComponent = Stick;
            return;
          case 'text':
            ChildComponent = TextBubble;
            break;
          default:
            break;
        }
        return ChildComponent && <ChildComponent {...(child as TextBubbleProps)} key={`stickchild-${idx}`} />;
      })}
      {/* </div> */}
    </div>
  );
};

export default chainClickable(Stick);
