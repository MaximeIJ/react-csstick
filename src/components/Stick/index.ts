export {default} from './Stick';
export * as StickPresets from './presets';
