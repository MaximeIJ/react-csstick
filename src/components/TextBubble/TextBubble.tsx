import React, {FC} from 'react';

import {chainCall, chainClickable} from '../../hooks/chainClickable';

import {Default} from './presets';

import './style.css';
import {stickCSSProps, textCSSProps} from '@/util/css';
import {TextBubbleProps} from '@/util/types';

const TextBubble: FC<TextBubbleProps> = (props = Default) => {
  const {font, fontSize, fontWeight, borderColor, borderRadius, color, dimensions, coord, text, onClick} = props;
  const {width, height, thickness} = {...Default.dimensions, ...dimensions};

  const stickStyle = stickCSSProps({
    width,
    height,
    thickness,
    color,
    coord,
  });
  const textStyle = textCSSProps({
    font,
    fontSize,
    fontWeight,
    borderColor,
    borderRadius,
  });
  const bubbleStyle = {
    ...stickStyle,
    ...textStyle,
  };

  return (
    <div className="bubble" style={bubbleStyle} onClick={chainCall(onClick)}>
      <div className="text">{text ?? ''}</div>
    </div>
  );
};

export default chainClickable(TextBubble);
