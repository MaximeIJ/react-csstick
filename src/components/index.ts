export * as Comic from './Comic';
export * as Scene from './Scene';
export * as Stick from './Stick';
export * as TextBubble from './TextBubble';
