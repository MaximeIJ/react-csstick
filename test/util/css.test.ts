import {describe, it} from 'vitest';

import {
  anglesCSSProps,
  arbitraryCSSProps,
  baseCSSProps,
  deg,
  divCss,
  multCss,
  offsetVar,
  rotate,
  rotateTransformCSSProp,
  stickCSSProps,
  textCSSProps,
} from '../../src/util/css';

import {StickPresets} from '@/components/Stick';
import {TextBubblePresets} from '@/components/TextBubble';

describe.concurrent('formatters', () => {
  it('multCss returns a css calc multiplying the expression passed', async ({expect}) => {
    expect(multCss('2px', 10)).toMatchSnapshot();
  });
  it('divCss returns a css calc divinding the expression passed', async ({expect}) => {
    expect(divCss('3px', 20)).toMatchSnapshot();
    expect(divCss('7px', 0)).toMatchSnapshot();
  });
  it('offsetVar returns a css calc offsetting the var passed from 100%', async ({expect}) => {
    expect(offsetVar('varName')).toMatchSnapshot();
  });
  it('deg returns a css deg of the optional number, or 0', async ({expect}) => {
    expect(deg(20)).toMatchSnapshot();
    expect(deg()).toMatchSnapshot();
  });
  it('rotate returns a css rotate of an optional angle, or 0', async ({expect}) => {
    expect(rotate(13)).toMatchSnapshot();
    expect(rotate()).toMatchSnapshot();
  });
});

describe.concurrent('CSS props', () => {
  it('rotateTransformCSSProp returns a css prop with a transform attribute for the optional value', async ({
    expect,
  }) => {
    expect(rotateTransformCSSProp(89)).toMatchSnapshot();
    expect(rotateTransformCSSProp()).toMatchSnapshot();
  });
  it('baseCSSProps returns optional css variables', async ({expect}) => {
    expect(baseCSSProps({})).toMatchSnapshot();
    expect(
      baseCSSProps({bgColor: '#123', color: '#456', width: '2vmin', height: '1vmax', thickness: '1pt'}),
    ).toMatchSnapshot();
  });
  it('anglesCSSProps returns css variables for one or more angles', async ({expect}) => {
    expect(anglesCSSProps(1, 'single')).toMatchSnapshot();
    expect(anglesCSSProps([91, 6, 10], 'multi')).toMatchSnapshot();
  });
  it('stickCSSProps returns css variables for optional stick props', async ({expect}) => {
    expect(stickCSSProps({})).toMatchSnapshot();
    expect(stickCSSProps(StickPresets.Default)).toMatchSnapshot();
    expect(
      stickCSSProps({...StickPresets.Default, coord: {x: '10%', y: '1px'}, offsets: {x: '10%', y: '1px'}}),
    ).toMatchSnapshot();
  });
  it('textCSSProps returns css variables for optional text props', async ({expect}) => {
    expect(textCSSProps({})).toMatchSnapshot();
    expect(textCSSProps(TextBubblePresets.Default)).toMatchSnapshot();
  });
  it('arbitraryCSSProps returns arbitrary css variables', async ({expect}) => {
    // Trivial, coverage
    expect(arbitraryCSSProps({})).toMatchSnapshot();
    expect(arbitraryCSSProps({a: 'b', c: 'd'})).toMatchSnapshot();
  });
});
