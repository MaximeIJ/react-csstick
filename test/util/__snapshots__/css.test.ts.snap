// Vitest Snapshot v1

exports[`CSS props > anglesCSSProps returns css variables for one or more angles 1`] = `
{
  "--single-0": "1deg",
}
`;

exports[`CSS props > anglesCSSProps returns css variables for one or more angles 2`] = `
{
  "--multi-0": "91deg",
  "--multi-1": "6deg",
  "--multi-2": "10deg",
}
`;

exports[`CSS props > arbitraryCSSProps returns arbitrary css variables 1`] = `{}`;

exports[`CSS props > arbitraryCSSProps returns arbitrary css variables 2`] = `
{
  "a": "b",
  "c": "d",
}
`;

exports[`CSS props > baseCSSProps returns optional css variables 1`] = `
{
  "--bgc": undefined,
  "--c": undefined,
  "--h": undefined,
  "--t": undefined,
  "--w": undefined,
}
`;

exports[`CSS props > baseCSSProps returns optional css variables 2`] = `
{
  "--bgc": "#123",
  "--c": "#456",
  "--h": "1vmax",
  "--t": "1pt",
  "--w": "2vmin",
}
`;

exports[`CSS props > rotateTransformCSSProp returns a css prop with a transform attribute for the optional value 1`] = `
{
  "transform": "rotate(89deg)",
}
`;

exports[`CSS props > rotateTransformCSSProp returns a css prop with a transform attribute for the optional value 2`] = `
{
  "transform": "rotate(0deg)",
}
`;

exports[`CSS props > stickCSSProps returns css variables for optional stick props 1`] = `
{
  "--base-0": "0deg",
  "--bgc": undefined,
  "--c": undefined,
  "--h": undefined,
  "--t": undefined,
  "--w": undefined,
  "left": "calc(0 - 0%)",
  "top": "calc(0 - 0%)",
}
`;

exports[`CSS props > stickCSSProps returns css variables for optional stick props 2`] = `
{
  "--base-0": "0deg",
  "--bgc": "transparent",
  "--c": "#888888",
  "--h": undefined,
  "--t": undefined,
  "--w": undefined,
  "left": "calc(0 - 0%)",
  "top": "calc(0 - 0%)",
}
`;

exports[`CSS props > stickCSSProps returns css variables for optional stick props 3`] = `
{
  "--base-0": "0deg",
  "--bgc": "transparent",
  "--c": "#888888",
  "--h": undefined,
  "--t": undefined,
  "--w": undefined,
  "left": "calc(10% - 10%)",
  "top": "calc(1px - 1px)",
}
`;

exports[`CSS props > textCSSProps returns css variables for optional text props 1`] = `
{
  "--bc": undefined,
  "--br": undefined,
  "--f": undefined,
  "--fs": undefined,
  "--fw": undefined,
}
`;

exports[`CSS props > textCSSProps returns css variables for optional text props 2`] = `
{
  "--bc": "transparent",
  "--br": "40%",
  "--f": "Comic Sans MS, Comic Sans, cursive",
  "--fs": "1.5rem",
  "--fw": "normal",
}
`;

exports[`formatters > rotate returns a css rotate of an optional angle, or 0 1`] = `"rotate(13deg)"`;

exports[`formatters > rotate returns a css rotate of an optional angle, or 0 2`] = `"rotate(0deg)"`;
